shotover-proxy/src/transforms/redis/cache.rs

```diff
@@ -74,386 +296,493 @@ impl SimpleRedisCache {
-                    }
+        CassandraStatement::Insert(insert) => {
+            for (c_name, operand) in insert.get_value_map().into_iter() {
+                let column_name = c_name.to_lowercase();
+                let relation_element = RelationElement {
+                    obj: Operand::Column(column_name.clone()),
+                    oper: RelationOperator::Equal,
+                    value: operand.clone(),
+                };
+                let value = value_map.get_mut(column_name.as_str());
+                if let Some(vec) = value {
+                    vec.push(relation_element)
                 } else {
-                    None
-                }
+                    value_map.insert(column_name, vec![relation_element]);
+                };
             }
-            _ => None,
+            Ok((
+                build_query_redis_key_from_value_map(
+                    table_cache_schema,
+                    &value_map,
+                    &insert.table_name.to_string(),
+                )?,
+                Bytes::new(),
+            ))
         }
-    } else {
-        None
+        CassandraStatement::Update(update) => {
+            populate_value_map_from_where_clause(&mut value_map, &update.where_clause);
+            Ok((
+                build_query_redis_key_from_value_map(
+                    table_cache_schema,
+                    &value_map,
+                    &update.table_name.to_string(),
+                )?,
+                Bytes::new(),
+            ))
+        }
+        _ => unreachable!(
+            "{} should not be passed to build_redis_key_from_cql3",
+            statement
+        ),
     }
 }
 
 #[async_trait]
 impl Transform for SimpleRedisCache {
+    async fn transform<'a>(&'a mut self, mut message_wrapper: Wrapper<'a>) -> ChainResponse {
+        let mut read_cache = true;
+        for m in &mut message_wrapper.messages {
+            if let Some(Frame::Cassandra(CassandraFrame {
+                operation: CassandraOperation::Query { query, .. },
+                ..
+            })) = m.frame()
+            {
+                for statement in &query.statements {
+                    debug!("cache transform processing {}", statement);
+                    match cql_statement::get_query_type(statement) {
+                        QueryType::Read => {}
+                        QueryType::Write => read_cache = false,
+                        QueryType::ReadWrite => read_cache = false,
+                        QueryType::SchemaChange => read_cache = false,
+                        QueryType::PubSubMessage => {}
+                    }
+                }
+            } else {
+                read_cache = false;
+            }
+        }
+        debug!("cache transform read_cache:{} ", read_cache);
+
+        // If there are no write queries (all queries are reads) we can read the cache
+        if read_cache {
+            match self.read_from_cache(message_wrapper.messages.clone()).await {
```

Comment:
Determining if we should read from the cache should be done per message.
Otherwise an unrelated write message would prevent reading from the cache.
I think this would happen really often and would make the cache near useless.

Reply:
Thinking about it more, it wouldnt be too terrible as it would require a single client to send a read/write in quick succession to hit this issue.
(I originally forgot that we are only processing messages from a single client here.)
But it still seems like an undesirable property and it would worsen if shotover was under high load.